Profile

friej715 opened the pull request:
![screen shot 2016-12-14 at 5 08 19 pm](https://cloud.githubusercontent.com/assets/1028403/21203506/c38c4df6-c220-11e6-8399-b88e036dbe8c.png)


Beginnings of profile page/starting work on #30. Trying to blow through issues before WEF.

What this includes:
* The ability to select and change almost all the basic demographics: age, gender, race, religion.
* Buttons read user data, and appropriate buttons are marked active on load.
* Success/error messages will occur on successful/unsuccessful `POST`.

What this doesn't include:
* There's no "Learn More" buttons. (Each question should have one, to explain why we're asking about this specific thing.)
* The "Read More" button in the beginning isn't wired up. (It should display our spiel about why we ask in general.)
* There's no selector for country. (Have to figure out how that's going to work!)

We should also quickly chat about the export data/delete account options and whether we're going to let people do that during the alpha, or if we're going to have them just email us.

friej715 committed:
profile page

friej715 committed:
cleaning up

friej715 committed:
formatting

friej715 commented:
Let's pause review on this 'til I get the country selector in.

chanderson0 commented on floodwatch/src/js/App.js:
```diff
@@ -43,6 +44,7 @@ export class App extends Component {
           <Route path="person" component={Person} onEnter={requireAuth} />
           <Route path="faq" component={Faq} />
           <Route path="compare" component={Compare} />
+          <Route path="profile" component={Profile} onEnter={requireAuth} />
```
Doesn't Compare need requireAuth also? I guess we need to clean this whole thing up... maybe let's do that separately after this is merged

chanderson0 commented on floodwatch/src/js/api/api.js:
```diff
@@ -188,6 +191,10 @@ export class FWApiClient extends APIClient {
     return res;
   }
 
+  async updatePersonDemographics(options: PersonDemographicRequest): Promise<void> {
```
As an FYI, the `async` annotation here is unnecessary since you're returning a `Promise` directly. You can leave it, there's no problem.

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+  constructor() {
+    super();
+    this.state = {
+      toPick: ['birth_year', 'twofishes_id', 'demographic_ids'],
```
This should probably not be in the individual component's state, but instead be a constant at the top of the file or somewhere else. It's not specific to the component, unless you have plans for it in the future...

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+    let filteredUserData = _.pick(UserData, this.state.toPick)
+
+    this.setState({
+      userData: filteredUserData
```
If we're going to use `PersonDemographicsRequest` to represent the resting state of the data as well, can we rename it to just `PersonDemographics`?

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+    })
+
+    if (event.target.name != 'age' && event.target.name != 'country_code') {
+      console.log(checked)
```
Remove console.log

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+
+  handleClick(checked: boolean, event: any): void {
+    let demo = event.target.textContent;
+    let demo_id = _.find(DemographicKeys.demographic_keys, (o: DemographicDictionary) => {
```
Feel free to leave this, but there's a shorthand for this too: `_.find(DemographicKeys.demographic_keys, { name: demo })` should in theory work

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+  render() {
+    let elems;
+    if (this.props.userData) {
+      if (this.props.filter.name == 'age') {
```
It feels like this should be in the calling instance... like you'd have a separate `Component` for the age selector. That would make the way the `DemographicContainer` render function works a little less JSON configurable, but it's not clear to me that it should be. For instance, you have `DemographicOption` taking both `updateYear` and `onClick` as props when you only ever need one. And since the country selector will be so different, it makes sense to put it in a different component.

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+          }
+
+          return <div key={key} className="custom-option">
+                      <Button href="#" 
```
Does button need an href?

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+type StateType = {
+  userData?: PersonDemographicRequest,
+  toPick: Array<string>,
+  curStatus: string
```
This is a good place to use the "enum" type that comes with FlowType. You could annotate this as `curStatus: null | "success" | "error"` and then be guaranteed you don't set it to anything else later.

friej715 replied:
Good call. It's so weird to merge the C++ and JS parts of my brain...

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+    userData.demographic_ids.push(id)
+    this.setState({
+      userData: userData,
+      curStatus: (this.state.curStatus != 'error') ? '' : 'error'
```
Maybe you want a `clearError()` method that handles this so it's not in a bunch of places?

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+        <Col xs={12}>
+        <Button onClick={this.updateUserInfo.bind(this)} style={{backgroundColor:'lightblue', width:'100%'}}>Submit</Button>
+        <ListGroup style={{display: displayGroup, textAlign:'center'}}>
+          <ListGroupItem bsStyle="success" 
```
This might be cleaner as:

```jsx
{ this.state.curStatus == 'success' &&
  <ListGroupItem bsStyle="success">
    Successfully saved changes!
  </ListGroupItem> }
{ this.state.curStatus == 'error' &&
  <ListGroupItem bsStyle="danger">
    Error while trying to save changes. Please check your connection.
  </ListGroupItem> }
```

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,198 @@
+
+  async updateUserInfo() {
+    try {
+      const reply = await FWApiClient.get().updatePersonDemographics(this.state.userData)  
```
I'd suggest reloading the person data by hitting `getCurrentPerson` again... or we can send back that response with the reply from the API maybe.

friej715 replied:
Good call. If we do the latter, does that get changed in `floodwatch-server/webserver/user.go` or somewhere else?

chanderson0 replied:
Yeah... I can build and deploy that

chanderson0 replied:
Done

chanderson0 commented:
Whoops, forgot to leave a comment. Biggest thing is [here](https://github.com/O-C-R/floodwatch/pull/53/files/ae5115a0fa8ffa122f9a9c19d49591cb381450ae#diff-9324d01b5c520feac165a185d88164f2R151) - we probably want different components for different option types.

friej715 committed:
intermediate work to get country in

friej715 committed:
Merge branch 'develop' into profile

friej715 committed:
adding features to profile

friej715 commented:
<img width="1541" alt="screen shot 2016-12-15 at 3 58 32 pm" src="https://cloud.githubusercontent.com/assets/1028403/21241799/5d005bd0-c2df-11e6-966e-4e495c954ea5.png">

I think this PR addresses all the comments above. In any case, it adds the last few features for the profile (for the alpha):
* "Learn more" buttons are implemented. 
* "Read more" button is implemented.
* We have our country selector, using react-autocomplete (so don't forget to run `yarn`), and hitting the twofishes API endpoint.

chanderson0 commented on floodwatch/src/js/api/api.js:
```diff
@@ -94,7 +94,10 @@ export class APIClient {
   async postJSON(path: string, body?: Object): Promise<any> {
     const res = await this.post(path, JSON.stringify(body));
-    return res.json();
+    // if (res.status == 204) {
```
This should still be here, no?

friej715 replied:
blorf, took that out while i was trying to track down a bug and forgot to put it back

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,225 @@
+import type {PersonDemographics} from '../api/types';
+import {AgeOption, LocationOption, DefaultOption} from './ProfileOptions'
+
+const ToPick = ['birth_year', 'twofishes_id', 'demographic_ids']; // stripping out admin, timestamp, etc.--other things that are set on the backend
```
JS constant format is CAPS_UNDERSCORE so this should be TO_PICK

chanderson0 commented on floodwatch/src/sass/App.scss:
```diff
@@ -42,3 +42,7 @@ h1,h2,h3,h4,h5 {
   margin-bottom:20px;
 }
+
+.testing > *{
```
Remove

chanderson0 commented on floodwatch/src/js/components/ProfileOptions.js:
```diff
@@ -0,0 +1,262 @@
+
+          <br/>
+          <Button className="learn-more" bsSize="xsmall" onClick={this.toggleDescriptionVisibility.bind(this)}>Learn more</Button>
+          <p style={{
```
```jsx
{ this.state.isDescriptionOpen && <p><Well bsSize="small">{this.props.filter.why}</Well></p> }
```

chanderson0 commented on floodwatch/src/js/components/ProfileOptions.js:
```diff
@@ -0,0 +1,262 @@
+  render() {
+    let elems;
+    if (this.props.userData) {
+      if (this.props.filter.name == 'age') {
```
I don't think you need any of these if / else ifs since everything will be neither age nor country

chanderson0 commented on floodwatch/src/js/components/Profile.js:
```diff
@@ -0,0 +1,225 @@
+          <h3>My Profile</h3>
+          <p>Donate your data to help us discover discriminatory patterns in advertising, and reverse the power relationship between people and advertisers.</p>
+          <p>Wondering why your demographic data matters? <Button bsSize="xsmall" onClick={this.toggleDescriptionVisibility.bind(this)}>Learn more</Button></p>
+          <p style={{
```
```jsx
{ this.state.isDescriptionOpen &&
  <p>
    ...
  </p> }
```

chanderson0 commented on floodwatch/src/js/components/ProfileOptions.js:
```diff
@@ -0,0 +1,262 @@
+          </div>
+          <br/>
+          <Button className="learn-more" bsSize="xsmall" onClick={this.toggleDescriptionVisibility.bind(this)}>Learn more</Button>
+          <p style={{
```
Same syntax as the other `<p>` blocks

chanderson0 commented on floodwatch/src/js/components/ProfileOptions.js:
```diff
@@ -0,0 +1,262 @@
+          </div>
+          <br/>
+          <Button className="learn-more" bsSize="xsmall" onClick={this.toggleDescriptionVisibility.bind(this)}>Learn more</Button>
+          <p style={{
```
`<p>` tag style

chanderson0 reviewed (changes requested):
Cool that autocomplete is working! A couple more things...

friej715 committed:
syntax tweaks

friej715 commented:
:+1:
